Ctagging 76 integration

mverzett opened the pull request:
First version of CTagging implementation
Adding @pvmulder and @imarches .

Commit:
First version of charm tagging
    -- Squash of 15 commits to make cherry picking less painful
    -- Skip commits that include files in the release
fix merging conflicts due to porting to 7.6

Commit:
added C vs B discriminator and automatic name extraction form XML files

Signed-off-by: Mauro Verzetti <mverzett@cern.ch>

Commit:
Final version of the charm tagging (with the TMVA reader)
--Squash of 10 commits to make cherry picking less painful
--removes all xml weight files from the commit history

Solved conflicts:
	PhysicsTools/PatAlgos/python/producersLayer1/jetProducer_cfi.py
	PhysicsTools/PatAlgos/python/recoLayer0/bTagging_cff.py
	PhysicsTools/PatAlgos/python/tools/jetTools.py
	RecoBTag/SecondaryVertex/python/secondaryVertex_cff.py

Commit:
adding testing (hardcoded)

Commit:
Swtich to GBRForest, use automatically generated hardcoded variable list for the variables to be used, provide code-generating script

Commit:
refactor sequences to run without exceptions on PAT with unscheduled mode

Commit:
Make code compatible for future condDB migration

cmsbuild commented:
A new Pull Request was created by @mverzett (Mauro Verzetti) for CMSSW_7_6_X.

Ctagging 76 integration

It involves the following packages:

DataFormats/BTauReco
PhysicsTools/PatAlgos
RecoBTag/CTagging
RecoBTag/Configuration
RecoBTag/SecondaryVertex
RecoBTau/JetTagComputer
RecoVertex/AdaptiveVertexFinder


The following packages do not have a category, yet:

RecoBTag/CTagging

@cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please review it and eventually sign? Thanks.
@TaiSakuma, @jdolen, @imarches, @makortel, @pvmulder, @acaudron, @mmarionncern, @GiacomoSguazzoni, @rappoccio, @rovere, @VinInn, @ahinzmann, @nhanvtran, @schoef, @ferencek, @gpetruc, @mariadalfonso, @cerati, @dgulhan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: bbe26d1130e715db419c80694bd0f7ee171a36e9
I found errors in the following unit tests: 
 
 ---> test runtestTqafTopTools had ERRORS
---> test runtestTqafTopEventProducers had ERRORS
---> test runtestTqafTopKinFitter had ERRORS
---> test runtestTqafTopHitFit had ERRORS
---> test runtestTqafTopJetCombination had ERRORS
---> test runtestTqafTopEventSelection had ERRORS
---> test runtestTqafExamples had ERRORS
---> test runtestPhysicsToolsPatAlgos had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/7923/summary.html

mverzett commented:
@degano, @slava77 
 this is really weird I tested it in CMSSW_7_6_X_2015-08-25-1100 and the matrix (```runTheMatrix.py -l limited -i all```) and unittests (```scram b -r -j 12 runtests``` ) were running smoothly (logs [here](https://mverzett.web.cern.ch/mverzett/ctagdev/integration/runTheMatrix.log
 ) and [here](https://mverzett.web.cern.ch/mverzett/ctagdev/integration/unittests.log
 )).

I see that now this build is not available anymore.

It looks like it's most likely a configuration problem. I still have to get the hang on the unscheduled running mode.

slava77 commented:
Hi Mauro,

I suggest you start from a fresh release and pick up the code as it's known to github (``` git cms-merge-topic 11122 ```)
I suspect some incomplete commits are in your test area.

ferencek commented on DataFormats/BTauReco/interface/TemplatedSoftLeptonTagInfo.h:
```diff
@@ -163,8 +163,8 @@ TaggingVariableList TemplatedSoftLeptonTagInfo<REF>::taggingVariables(void) cons
     list.insert( TaggingVariable(btau::leptonQuality2, data.quality(SoftLeptonProperties::Quality::btagLeptonCands, false)), true );
-    list.insert( TaggingVariable(btau::trackSip2dSig,  data.sip2d),    true );
-    list.insert( TaggingVariable(btau::trackSip3dSig,  data.sip3d),    true );
+    list.insert( TaggingVariable(btau::trackSip2dSig,  data.sip2d),    true ); //These are correct! it turns out is just a bad naming convention
```
Please remove these comments. This (very old) inconsistency will be addressed in a separate PR (as announced in https://twiki.cern.ch/twiki/bin/view/CMS/CMSSW_7_6_0).

mverzett replied:
OK, I will remove the comments, but please be aware that data.sip[23]d do not map to the value but to the significance. This means that the TaggingVariableName is correct, is the object data member to be misleading

ferencek commented on DataFormats/BTauReco/src/TaggingVariable.cc:
```diff
@@ -127,7 +127,7 @@ const char* const TaggingVariableTokens[] = {
   /* [trackPParRatio]                           = */ "trackPParRatio",
 
-  /* [trackIp2dVal]                             = */ "trackIp2dVal",
+  /* [trackIp2dSig]                             = */ "trackIp2dSig",
```
Removing old tagging variables could lead to backward incompatibilities. Please only add new variables. However, a bigger problem is that this change is inconsistent. Why would you just change the token but not the corresponding description. Also, where are the corresponding changes in the header file?

mverzett replied:
I know, before sending this PR I discussed this in a BTV software meeting with @imarches, @pvmulder, and @acaudron, and outlined the problem in [this](https://indico.cern.ch/event/441479/session/0/contribution/8/attachments/1145703/1642478/BTV_CTagDev_27Aug2015.pdf) presentation (slide 9).
Also, if you look in [TaggingVariable.h](https://github.com/mverzett/cmssw/blob/ctagging_76Integration/DataFormats/BTauReco/interface/TaggingVariable.h) you will notice that "trackIp2dVal" never existed, and that name was a bug.

I checked by running the full BTagging sequence that the commit was non-breaking before pushing it in anyways. 

For the future, always in that presentation, I suggested to move the TaggingVariableToken to be a (unordered_)map to make this sort of mistake more difficult to make. What do you think about it?

ferencek replied:
Sorry, Mauro, I was a bit too quick with my comment. Yes, you are completely right, there is indeed an inconsistency in the TaggingVariable code. I suggest you simply remove all instances of either `trackIp2dVal` or `trackIp2dVal` since neither of them seems to be used anywhere. Also, can you also rename `trackJetDist` to `trackJetDistVal` in `TaggingVariableTokens`.

Concerning the unordered map, I guess it will require a long initialization list in the constructor where the same type of mistakes and typos can still happen. So unless that would speed up the code, I would not bother with redesigning this part of the code.

mverzett replied:
OK, once I manage to have some better connection I will make the checks and the changes you suggested.

Concerning the map I agree that typos can still happen, but the person writing the code will not have to match the token position with the enum position, and since c++11 will have ```{TaggingVariableName, TaggingVariableToken}``` pairs that, in my opinion, are easier to handle. It should also slightly speed up the code. So far the Token to Name matching is done O(N), with a map it would be done in O(LogN), with an unordered map in O(1), but the memory footprint would be larger and the constant time might be longer than LogN for few entries.

makortel replied:
I know ~nothing about the context, but want to comment about the map (as in the linked presentation) if it has not been discussed yet (if it has, sorry for the noise). The `map<const char *, T>` in slide 10 might not be what you want, as `operator[]/find` will compare the pointer, not the contents of the C-string. If you wanto to compare the C-string contents, you need to add a custom comparator to the map template parameters. And since we're talking about performance, you might also want to test sorted vector of pairs + binary search, which is often faster than similar map for searching.

ferencek replied:
@makortel, thanks for pointing this out. Indeed, it might be better to use `std::string` as the key type but then all of the downstream code needs to be appropriately updated. Also note that we would not be able to use `operator[]` to access the map elements, since the map needs to be `const`. Instead, map elements would have to be accessed using `at()`.

@mverzett, not being a map expert, I'm reading online that `map` vs `unordered_map` performance can depend on the map size and the exact implementation of the `unordered_map` so to me it is not clear what would work best in this case. However, I do agree that not having to worry about the variable order would make the code less error prone so using maps should generally be better than what is currently being used. Hence, if you feel like implementing maps, go for it :)

VinInn replied:
do not use maps, do not use lists unless you have an algorithm that makes an optimal use of their peculiar data structures. They are very memory greedy and inefficient in particular for write once read many cases. 
if possible use a (sorted) std::vector.   unordered_map is also a possibility

ferencek replied:
From what I read, `std::unordered_map` is even worse than `std::map` in terms of memory usage but provides faster insertion and retrieval time.

mverzett replied:
I understood the same as Dinko from what I had read. But also consider that:

  * these changes will not happen in this PR (maybe not even in this release)
  * the map can be const, static, and shared across the threads, so the the memory usage is pretty limited.

mverzett replied:
@VinInn , for my personal knowledge, can you detail a bit more the statement of maps beeing memory greedy and inefficient for write once-read many tasks? Thank you.

VinInn replied:
google for "map vs vector C++" and you will find all possible arguments (unfortunately many quite confusing)
essentially nodes in maps are allocated  on the heap individually and have memory overhead (list list) due to pointers between nodes.
vector has no memory overhead and has contiguous storage.
for simple payloads (float, int, small struct) it is pretty obvious that binary_search in a sorted vector will be faster
than map::find and will use much less space...
For larger payload caching issues will play a role... 
For string is even more complex as string are stored on the heap and therefore there is already the cost of sparse storage and cache misses
For string in particular hashing is very efficient: so unorder_map is most probably the best choice.
Actually for a finite (i..e known in advance) number of strings one can use 64bit hashing, verify that is unique and then use a double vector (ordered hashes and corresponding string),  search the hashing return the string at the corresponding location.
This is of course an over-killing for your application...
So just use unordered_map and verify that is NOT a bottle-neck

ferencek replied:
Please remove any appearance of `trackIp2dVal` and `trackIp2dSig` in `DataFormats/BTauReco/interface/TaggingVariable.h` and `DataFormats/BTauReco/src/TaggingVariable.cc`. These are not used anywhere so better to remove them.

mverzett commented:
Hi @slava77, at the moment I am in the CMSWeek with sub-par connectivity, to say the least. I will try these changes ASAP.

slava77 commented:
On 9/7/15 4:34 PM, Mauro Verzetti wrote:
> Hi @slava77 <https://github.com/slava77>, at the moment I am in the
> CMSWeek with sub-par connectivity, to say the least. I will try these
> changes ASAP.

OK.
I don't think we are in a hurry, if the target is just 76X.


> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/11122#issuecomment-138381047>.
>

mverzett commented:
Indeed it seems that something went lost when packing the commits for the PR and removing the xml weight files from the log history.

I will work for fixing it ASAP

slava77 commented on RecoBTag/CTagging/interface/CharmTagger.h:
```diff
@@ -0,0 +1,63 @@
+#include "RecoBTau/JetTagComputer/interface/JetTagComputerRecord.h"
+
+//DEBUGGING
+#include "TFile.h"
```
please remove the TFile and TNtuple
If really needed for debugging, hide them under ifdefs to be picked up by recompilation

Dr15Jones replied:
Also, there is no way TFile nor TNtuple can be made thread safe. These must be removed.

slava77 commented on RecoBTag/CTagging/interface/CharmTagger.h:
```diff
@@ -0,0 +1,63 @@
+
+private:
+  mutable std::mutex mutex_;
+	[[cms::thread_guard("mutex_")]] std::unique_ptr<TMVAEvaluator> mvaID_;
```
Can we move to GBRForest as done in many other taggers?

slava77 replied:
... actually, if you are using GBRForest (seeing below that it's happening), this mutex is not needed (just looking at the recently made changes e.g. in soft lepton taggers in #10543 )

VinInn replied:
We SHALL move to GBRForest for any new code (and migrate fast any old one!)
TMVAEvaluator shall not be allowed anymore in production code

mverzett replied:
TMVAEvaluator ***does*** use the GBRForest implementati under the hood, but still allows for using plain TMVA.  I left the option to choose from the configuration. Given that I thought that it was safer to leave the mutex. For consistency I could remove the option and fix to GBRForest and remove the mutex.

BTW, for my own learning, what does the syntax ```  [[cms::thread_guard("mutex_")]]  ``` mean? I looked it up on Google but I did not quite understand the effect.

Dr15Jones replied:
```[[ ]]``` is C++11's new attribute syntax. An attribute is just a way to pass implementation specific information to the compiler. The static analyzer we use is based on the clang compiler and we have extended it to look for attributes which begin with ```cms::```. In this case the ```[[cms::thread_guard("mutex_")]]``` tells the static analyzer that the variable declared on that line is thread safe because it is protected from being called by multiple threads simultaneously by the variable named ```mutex_```.

Dr15Jones replied:
TMVAEvaluator has its own internal ```std::mutex```
http://cmslxr.fnal.gov/lxr/source/CommonTools/Utils/interface/TMVAEvaluator.h?v=CMSSW_7_6_X_2015-09-08-2300#0039
therefore you do not need a mutex here.

slava77 commented on RecoBTag/CTagging/interface/CharmTagger.h:
```diff
@@ -0,0 +1,63 @@
+  /// explicit ctor 
+	CharmTagger(const edm::ParameterSet & );
+	~CharmTagger();//{}
+  virtual float discriminator(const TagInfoHelper & tagInfo) const override;
```
please use more consistent indentation (this refers to all of the new code)

slava77 commented on RecoBTag/CTagging/python/RecoCTagging_cff.py:
```diff
@@ -0,0 +1,17 @@
+from RecoBTag.CTagging.pfInclusiveSecondaryVertexFinderCvsLTagInfos_cfi import *
```
is this file really needed?
can we define the sequence in cTagging_cff ?

mverzett replied:
For some reason I do not fully understand, I need the two definitions to be separate. Otherwise the default tests running in unscheduled mode will fail. The PAT sequence will try to re-build the TagInfos instead of taking what is provided by the AOD event content

ferencek replied:
One of the features of the unscheduled mode is that it will run any loaded modules whose names match the required input products. For b-tagging in particular, this essentially means that when PAT is run, for example, to produce MiniAOD, all required b-tag discriminators will be re-run and not taken from RECO/AOD. This is good for cases where one wants to run the latest version of the taggers and ignore discriminators already stored in the event.

So what I don't understand is why would the PAT tests fail. Re-building TagInfos from AOD should not cause any failures.

slava77 commented on RecoBTag/CTagging/python/candidateNegativeCombinedSecondaryVertexSoftLeptonCvsLComputer_cfi.py:
```diff
@@ -0,0 +1,11 @@
+
+from RecoBTag.CTagging.candidateCombinedSecondaryVertexSoftLeptonCvsLComputer_cfi import *
+
+candidateNegativeCombinedSecondaryVertexSoftLeptonCvsLComputer = candidateCombinedSecondaryVertexSoftLeptonCvsLComputer.clone()
```
you will use less letters if the modified values are passed as arguments in the clone method
```python
candidateNegativeCombinedSecondaryVertexSoftLeptonCvsLComputer = candidateCombinedSecondaryVertexSoftLeptonCvsLComputer.clone(
    vertexFlip = True,
    trackFlip = True,
....
)
```

slava77 commented on RecoBTag/CTagging/python/charmTagJetTags_cfi.py:
```diff
@@ -0,0 +1,17 @@
+import FWCore.ParameterSet.Config as cms
+
+#pfCombinedSecondaryVertexSoftLeptonCvsLJetTags 
```
is this commented out line needed?

slava77 commented on RecoBTag/CTagging/python/charmTaggerProducer_cfi.py:
```diff
@@ -0,0 +1,40 @@
+   mvaName = cms.string('BDT')
+   )
+
+## charmTagsComputerCvsL.variables = cms.VPSet( 
```
remove the commented out lines, please, unless there is a good reason to keep

slava77 commented on RecoBTag/CTagging/src/CharmTagger.cc:
```diff
@@ -0,0 +1,149 @@
+#include <map>
+#include <iostream>
+
+#include "TDirectory.h" //DEBUG
```
drop or hide with ifdef from the regular build

slava77 commented on RecoBTag/CTagging/src/CharmTagger.cc:
```diff
@@ -0,0 +1,149 @@
+	sl_computer_(configuration.getParameter<edm::ParameterSet>("slComputerCfg")),
+	vars_definition_(configuration.getParameter<vpset>("variables")),
+  mva_name_( configuration.getParameter<std::string >("mvaName") ),
+  use_condDB_(configuration.existsAs<bool>("useCondDB") ? configuration.getParameter<bool>("useCondDB") : false),
```
existsAs usually should not be used, the default values should be specified in the fillDescriptions method.
We made exceptions in the past for old code to keep adding new parameters with existsAs.
This is a new code and we should have fillDescriptions defined for it

slava77 replied:
oh, well, the fillDescriptions implementation is not that easy, I guess; if it needs to be propagated to the btaggers framework

VinInn replied:
It is not difficult, just a bit of additional typing work.
It shall be propagated to the btaggers framework, please try.
If any showstopper shall appear, better we know.

slava77 commented on RecoBTag/CTagging/src/CharmTagger.cc:
```diff
@@ -0,0 +1,149 @@
+  }
+	
+  //DEBUG
+	if(debug_mode_){
```
ifdef BTAG_DEBUG or something like that to not compile by default

slava77 commented on RecoBTag/CTagging/src/CharmTagger.cc:
```diff
@@ -0,0 +1,149 @@
+float CharmTagger::discriminator(const TagInfoHelper & tagInfo) const {
+  // default value, used if there are no leptons associated to this jet
+  const reco::CandIPTagInfo & ip_info = tagInfo.get<reco::CandIPTagInfo>(0);//"pfImpactParameterTagInfos");
+	const reco::CandSecondaryVertexTagInfo & sv_info = tagInfo.get<reco::CandSecondaryVertexTagInfo>(1);//"pfInclusiveSecondaryVertexFinderCtagLTagInfos");
```
please use consistent indentation

slava77 commented on RecoBTag/Configuration/python/RecoBTag_EventContent_cff.py:
```diff
@@ -41,9 +40,14 @@
-##	'keep *_pfCombinedSecondaryVertexSoftLeptonCtagLJetTags_*_*',
-        'keep *_inclusiveCandidateSecondaryVertices_*_*'
+        'keep *_pfCombinedSecondaryVertexSoftLeptonBJetTags_*_*',
+        ##	'keep *_pfCombinedSecondaryVertexSoftLeptonCtagLJetTags_*_*',
```
is this still needed?
Somehow I understood that the "new" naming convention is Cvs*

slava77 commented on RecoBTag/Configuration/python/RecoBTag_EventContent_cff.py:
```diff
@@ -87,9 +91,13 @@
-##	'keep *_pfCombinedSecondaryVertexSoftLeptonCtagLJetTags_*_*',
-        'keep *_inclusiveCandidateSecondaryVertices_*_*'
+        'keep *_pfCombinedSecondaryVertexSoftLeptonBJetTags_*_*',
+        ##	'keep *_pfCombinedSecondaryVertexSoftLeptonCtagLJetTags_*_*',
```
I think thought these were commented out just until ctagging code is in place. Should these be uncommented now?

mverzett commented:
Hi @slava77 , thanks for the comments. Starting from Monday I will go through them and fix everything. The indentation I guess was hidden by the editor. And some naming inconsistency is due, as I said, to messing up with git trying to remove the xml weights form the history.

mverzett commented:
Hi @slava77 , I was at the point of packing two new commits to add to the PR to become compliant to most of the suggestions in this thread, but when running ```git status``` I get back: ```fatal: unable to read tree 7d7f3a625e45833c47c08cfcb4ac86e74a8a267d```. And nothing else.

Do you (or anyone else), know what is the source of this issue? And how to solve it? I have been running in the work/ area of lxplus.

Thanks

Mauro

slava77 commented:
@smuzaffar @degano do you know about the last issue?
I suspect it's some changes on the git cache side.
I've seen the problem before.

smuzaffar commented:
@slava77 , I really do not know how to solve it. I would have tried to fix it by doing something like
- creating a new-cmssw-dev-area
- cmsenv
- git cms-addpkg <packages ) I want to update
- copy files from current work dir to my new-cmssw-dev-area
this will lose your current 8 commits which you have in this PR though

mverzett commented:
I was thinking to do the following:
   * Set-up a new dev area
   * merge-topic this PR
   * Run a big diff of all the packages to get the commits I have done. Make a patch
   * apply the patch to the new dev area

If there is no better solution I will go for this one

smuzaffar commented:
@mverzett , yes that should work and even keep the history of your commits in this PR

slava77 commented:
On 9/15/15 7:13 AM, Malik Shahzad Muzaffar wrote:
> @slava77 <https://github.com/slava77> , I really do not know how to
> solve it. I would have tried to fix it by doing something like
> 
>   * creating a new-cmssw-dev-area
>   * cmsenv
>   * git cms-addpkg <packages ) I want to update
>   * copy files from current work dir to my new-cmssw-dev-area this will
>     lose your current 8 commits which you have in this PR though
> 

Yes, the recovery will go this way or similar (start a new git area)

Still, what's the main reason for these issues to happen and how often
do they (or should they) be happening so that people can avoid them.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/11122#issuecomment-140369931>.
>

ferencek commented:
@mverzett, assuming your current development branch does not have any complicated merge commits, I would suggest the following steps:
```
cmsrel CMSSW_7_6_0_pre5
cd CMSSW_7_6_0_pre5/src
cmsenv
git fetch official-cmssw
git cms-merge-topic 11122 # this is just to check out all the affected and dependent packages
git fetch official-cmssw refs/pull/11122/head:ctagging_76Integration
git branch 760pre5_for_rebase CMSSW_7_6_0_pre5
git checkout ctagging_76Integration
git rebase 760pre5_for_rebase
```
If all goes well, you will have all the commits in this PR rebased to CMSSW_7_6_0_pre5.

Generally, for PRs it is better to avoid merge commits. They make the topology of the commits history more complicated than needed and there used to be cases of very confusing and hard to interpret GitHub diffs.

mverzett commented:
@ferencek, I will try to do it in this way. My only worry is that my last developments were based on CMSSW_7_6_X_2015-09-06-2300 to have all the latest changes. And I really hope that the diff will not pick up changes from other commits.

ferencek commented:
@mverzett, you actually have a merge commit (https://github.com/mverzett/cmssw/commit/eff3442ac5abb9239d90781f95bd5e43c1326908) on your development branch so I tried cherry-picking individual commits (along the way I had to manually resolve a few minor merge conflicts) and the final results is https://github.com/cms-btv-pog/cmssw/tree/ctagging_76Integration Note that CMSSW_7_6_0_pre5 was used as the base on top of which your commits were applied.

Commit:
Fix crash due to merge issues. Polish code, make it compliant to comments

mverzett commented:
Hi All,

I changed the code according to your suggestions. Changelog:
   * Fixed the crashes due to problems in merging/cherry picking of the previous dev branches
   * Removed TTree and TNtuple debugging from the code.
   * Removed the existsAs<> from the code. Now everything is required and has a default value provided in the cfi. I think the usage of fillDescriptors can be postponed to future PR, as it would refactor quite some code.
   * Fixed indentation in the code, removed unused comments. General polishing.

Tested with ```runTheMatrix.py -l limited -i all```, the log can be found (here)[https://mverzett.web.cern.ch/mverzett/ctagdev/matrix.log]. I did not run unittests but I expect some pat ones to crash. The PAT sequence needs some input tag infos currently missing in the AODs, this PR adds the necessary data to the event content.

Using @ferencek recipe to rebase the commit history looks cleaner, but I had to force the push in my topical branch. Please have a look if the forced push screwed up something.

Thank you

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

ferencek commented on DataFormats/BTauReco/interface/TemplatedSoftLeptonTagInfo.h:
```diff
@@ -163,8 +163,8 @@ TaggingVariableList TemplatedSoftLeptonTagInfo<REF>::taggingVariables(void) cons
-    list.insert( TaggingVariable(btau::trackSip2dSig,  data.sip2d),    true );
-    list.insert( TaggingVariable(btau::trackSip3dSig,  data.sip3d),    true );
+    list.insert( TaggingVariable(btau::trackSip2dSig,  data.sip2d),    true ); 
+    list.insert( TaggingVariable(btau::trackSip3dSig,  data.sip3d),    true ); 
```
Can you remove spaces or tabs at the end of these two lines? Otherwise, this will probably create unnecessary merge conflicts with/for #11247, depending on which PR gets merged first.

ferencek commented on PhysicsTools/PatAlgos/python/producersLayer1/jetProducer_cfi.py:
```diff
@@ -47,7 +47,12 @@
+        cms.InputTag("pfCombinedMVABJetTags"),
+        #CTagging
+        cms.InputTag('pfCombinedCvsLJetTags'),
+        cms.InputTag('pfCombinedCvsBJetTags')
```
To prevent PAT tests from failing, I suggest for now to keep the c-tagging additions in this cfi file commented out. Once RelVals with required inputs become available, this can be revisited.

mverzett replied:
OK, as long as miniAOD production is done afterwards

Commit:
apply @ferencek comments

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

ferencek commented on PhysicsTools/PatAlgos/python/producersLayer1/jetProducer_cfi.py:
```diff
@@ -47,7 +47,12 @@
         cms.InputTag("softPFElectronBJetTags"),
-        cms.InputTag("pfCombinedMVABJetTags")
+        cms.InputTag("pfCombinedMVABJetTags"),
+        cms.InputTag("pfCombinedSecondaryVertexSoftLeptonBJetTags"),
```
Was it discussed and agreed anywhere to add `pfCombinedSecondaryVertexSoftLeptonBJetTags` here? From the last performance studies I've seen, this tagger was not performing according to the expectation so unless that changed, I would rather leave out of the default set of taggers. Also, `pfCombinedMVABJetTags` now appears twice.

mverzett replied:
I guess this is something that was present in my previous build but then got removed and I mistakenly replaced. Fixing it.

Commit:
remove discriminator duplicate

slava77 commented:
Mauro, are you done with comments feedback so far?

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

mverzett commented:
I would say so, if no one else has anything to comment.

ferencek commented:
There is still this comment https://github.com/cms-sw/cmssw/pull/11122#discussion_r39549034

Commit:
remove trackIp2dVal/Sig from tagging variables

mverzett commented:
This should be all

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test
[tests will fail, AFAIK, because the IB is broken for runtime]

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 94d0189fc2fee5b9334ae8b5c8a0bc273a370576
I found errors in the following unit tests: 
 
 ---> test runtestPhysicsToolsPatAlgos had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8104/summary.html

ferencek commented on PhysicsTools/PatAlgos/test/patTuple_addBTagging_cfg.py:
```diff
@@ -104,6 +104,9 @@
     ,'pfPositiveCombinedMVABJetTags'
     ,'pfNegativeCombinedMVABJetTags'
+    #CTagging
+    ,'pfCombinedTMVACtagLJetTags'
```
Unit tests are failing due to unsupported tagger names. The names here should be changed to `pfCombinedCvsLJetTags` and `pfCombinedCvsBJetTags`.

Commit:
fix naming convention: Ctag[LB] --> Cvs[LB] everywhere where was missing

mverzett commented:
OK, it should be fixed now, I'm very sorry for the mess. While I was working on it I noticed that there were still some elements named with the old Ctag[LB] convention and I renamed them accordingly.

For my own knowledge: is there a way to run the very same tests that jenkins runs and possibly get back a summary page/log/comment as cmsbuild does? This would be very helpful. Going through endless logs I might miss something.

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test
[the IB is still bad though, so some failures in jenkins are expected]

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
 
Tested at: 2dd124dd3ba7886cf32cfb512468ca8411cb8ad4
I found errors in the following unit tests: 
 
 ---> test runtestPhysicsToolsPatAlgos had ERRORS
 
 
you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8117/summary.html

mverzett commented:
The error comes from disabling the CTagging discriminators from the possible PAT discriminators, while we wait for the RelVals to be done. What should I do? Should I comment out consistently the CTagging discriminators everywhere in the PAT code while we wait for RelVals to be ready or we leave it like that, given that running the matrix does not cause problems? (And we fix it once we get the RelVals by un-commenting the option)

ferencek commented:
@mverzett, thanks for updating and using a consistent naming convention. However, you haven't fully addressed my original comment, namely that the taggers added in `PhysicsTools/PatAlgos/test/patTuple_addBTagging_cfg.py` have `TMVA` in their names but were added without `TMVA` to ` PhysicsTools/PatAlgos/python/recoLayer0/bTagging_cff.py `. This is what is causing that particular PAT test to fail. Just remove `TMVA` from tagger names in `PhysicsTools/PatAlgos/test/patTuple_addBTagging_cfg.py` and this should fix the problem.

For running PAT tests, I usually go to `PhysicsTools/PatAlgos/test/` and run
```
./runAsyncTests.pl -a
```
Another option that @slava77 posted in #10197 is to run
```
scram b -r -j 12 runtests >& b.log
```

mverzett commented:
Hi @ferencek, if I recall correctly we decided to **add** TMVA in the name. Just to make clear which tagger it is, in case other training options appear. I will make everything consistent and run the tests this time before committing.

Commit:
minor bugfix on the test script

ferencek commented:
OK, as long as the naming convention is consistent. However, note that 
we have other TMVA-based taggers without TMVA in their name, and if 
eventually this particular tagger is switched to GBRForest, I would say 
there is no strong motivation to have TMVA in its name. What would maybe 
make more sense is to have BDT in the name but one could also argue 
against that if at some point we end up having more than one BDT-based 
tagger.

On 09/16/2015 12:17 PM, Mauro Verzetti wrote:
> Hi @ferencek <https://github.com/ferencek>, if I recall correctly we
> decided to *add* TMVA in the name. Just to make clear which tagger it
> is, in case other training options appear. I will make everything
> consistent and run the tests this time before committing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/11122#issuecomment-140792127>.
>

mverzett commented:
Hi, I removed the comment as I noticed that actually Petra asked to go in the **other** direction and that file was left there as it did not compromise any running sequence. It should be fixed now. I ran the tests and they worked. Hopefully everything should be in place now.

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8117/summary.html

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ee958d04b4b7ce211e39b34feb4bf547dab91523
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8138/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8138/summary.html

ferencek commented:
@mverzett, this PR now has merge conflicts after #11198 was merged. Maybe just rebase to the CMSSW_7_6_X head.

slava77 commented:
@mverzett 
how is that rebase going?

mverzett commented:
I had to start back from scratch, but I should have managed to merge. However I have not been able to test it as the compilation of the head crashes with this error:

```
RecoMET/METFilters/plugins/HcalStripHaloFilter.cc:38:52: error: 'const class reco::BeamHaloSummary' has no member named 'getProblematicStrips'
   auto const& problematicStrips = beamHaloSummary->getProblematicStrips();
```

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
did you do ``` git cms-checkdeps -a ``` ?
which IB had the problem?

Commit:
Merged refs/pull/11122/head from repository cms-sw

ferencek commented:
I suggest we just fire up the tests. I'm afraid that just running `git cms-checkdeps -a` will not necessarily resolve all the problems if Mauro just rebased his development branch to the 76X HEAD. For `cms-checkdeps` to work he would have to checkout all the packages which were modified since 760pre5, which I believe Mauro is using, and the 76X HEAD.

mverzett commented:
I am trying to run on the last integration build. But running the automated checks surely won't hurt.

slava77 commented:
where did you have conflicts?
I don't really want to start tests in jenkins until there is a new baseline with respect to which changes from this PR can be compared.
From the recently integrated PRs it seems to me that the conflict was with code from  #11247  that went to CMSSW_7_6_X_2015-09-18-1100

mverzett commented:
The conflicts were only located in https://github.com/cms-sw/cmssw/blob/CMSSW_7_6_X/DataFormats/BTauReco/interface/TemplatedSoftLeptonTagInfo.h#L170
As I left a FIXME comment and someone (I think @imarches ) fixed it. 

Tried the rebase on the head of 7.6.X but in 7.6.0.pre5, so I guess that was the reason it did not compile. 

I am compiling with the latest IB at the moment (CMSSW_7_6_X_2015-09-17-2300 )

ferencek commented:
@mverzett, are you sure the conflict was in `DataFormats/BTauReco/interface/TemplatedSoftLeptonTagInfo.h`? This file did not and still does not have modifications as part of this PR. However, you were modifying that file at some point in the history so that explains why you had conflicts during the rebase. But that's not where the actual merge conflicts were happening since they started appearing as soon as #11198 was merged, which was before #11247 was merged.

Since CMSSW_7_6_X_2015-09-18-1100 does not seem to be available, probably the best would be to rebase to CMSSW_7_6_X_2015-09-17-2300 and then there should be a proper baseline defined. Note that CMSSW_7_6_X_2015-09-17-2300 does not contain #11247 but as said above, I do not expect conflict with that PR but rather with #11198 which is contained in CMSSW_7_6_X_2015-09-17-2300.

mverzett commented:
Indeed it's probably it. As I said it is still compiling (which good, I did not get any issue so far)

mverzett commented:
Compilation just terminated. Merge topic seems to work smoothly on the IB. @slava77, can you please trigger the tests?

slava77 commented:
it's good that it compiles, but it will take until CMSSW_7_6_X_2015-09-18-2300 to get meaningful jenkins tests and comparisons.
Right now, if I start the tests I will get everything in this diff https://github.com/cms-sw/cmssw/compare/CMSSW_7_6_X_2015-09-17-2300...5d6fb10 added as well

mverzett commented:
Sure, no problem. I will wait for the results.


sent from my mobile, please excuse brevity and spelling errors

Sent using CloudMagic<https://cloudmagic.com/k/d/mailapp?ct=pa&cv=7.2.9&pv=5.1>
On ven, set 18, 2015 at 7:52 PM, Slava Krutelyov <notifications@github.com<mailto:notifications@github.com>> wrote:


it's good that it compiles, but it will take until CMSSW_7_6_X_2015-09-18-2300 to get meaningful jenkins tests and comparisons.
Right now, if I start the tests I will get everything in this diff CMSSW_7_6_X_2015-09-17-2300...5d6fb10<https://github.com/cms-sw/cmssw/compare/CMSSW_7_6_X_2015-09-17-2300...5d6fb10> added as well

—
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/11122#issuecomment-141519010>.

mverzett commented:
For the record: both PatAlgos unittest and Matrix test work in my local area.

slava77 commented:
@cmdbuild please test

mverzett commented:
@slava77 is there any news? I do not see jenkins checks being triggered.

slava77 commented:
I started the tests just now

cmsbuild commented:
+1
Tested at: 4ac77f290e46ed4ac234971fa2ca5bd5699ae6c8
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8257/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8257/summary.html

slava77 commented on RecoBTag/CTagging/BuildFile.xml:
```diff
@@ -0,0 +1,13 @@
+<use name="CommonTools/Utils"/>
+<use name="roottmva"/>
+<use name="FWCore/Utilities"/>
+<use name="DataFormats/BTauReco"/>
```
this generates the warning
```
***WARNING: Multiple usage of "DataFormats/BTauReco". Please cleanup "use" in "non-export" section of "src/RecoBTag/CTagging/BuildFile".

```

slava77 commented:
During compilation, I get 
```
***WARNING: Multiple usage of "DataFormats/BTauReco". Please cleanup "use" in "non-export" section of "src/RecoBTag/CTagging/BuildFile".
```

@mverzett please correct the file

@smuzaffar @degano can we add a check for scram warnings?

slava77 commented:
the warning can also be seen in the jenkins build
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8257/build.log

slava77 commented:
@mverzett what is the status of adding the ctaggers to DQM plots?

Commit:
fix scram warning

mverzett commented:
@slava77 fixed.

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

gvonsem commented:
@slava77 The code for 1D c-tag efficiency plots is ready, while some other features we (mainly Keng Kovitanggoon) are still working on, but progress is made. These include 
- putting the c-tag input variables in DQM
- including light rejection vs b rejection curves for constant c efficiency in DQM

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on RecoBTag/CTagging/src/CharmTagger.cc:
```diff
@@ -0,0 +1,106 @@
+		mva_var.index = mva_var.has_index ? var.getParameter<int>("idx") : 0;
+		mva_var.default_value = var.getParameter<double>("default");
+
+		variables_.push_back(mva_var);
```
vars_definition_ is used only to define variables_.
I think vars_definition_ should be eliminated and variables_ vector is to be filled in the constructor.

slava77 replied:
there are about 60kB allocated of fairly useless PSet copy in the vars_definition_

cmsbuild commented:
+1
Tested at: 599a3d194dc99089fa942deb95caf1855eb5b409
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8318/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8318/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/11122	599a3d1
- code changes are OK
    - cleanup of the locally stored VPSet can be done in a separate PR; although it would be nice if it happens in this PR
- jenkins tests pass and comparisons with baseline show no differences
- tested locally in CMSSW_7_6_X_2015-09-22-1500 /my test area sign593/ to check technical performance and add the new discriminants to the reco product fwlite validation 
    - new modules add about 0.3% to the processing time ( in 251721); this so far is within a balance of recent removal of the legacy taggers
    - pfCombinedCvsLJetTags and pfCombinedCvsBJetTags show up in RECO and AOD as expected; in addition RECO has a corresponding product of tag infos. The net increase per event is about 0.1% (using run2 data file)
     - allocated memory increases by about 1MB of these, 0.87 MB is from the GBRForest, and 0.13 from the VPSet of parameters (one copy held by the framework, and one copy held by the module itself); expanded configs grew by 1.3K lines (~50K chars) of python.

Looking at the new tags, the discriminants show up in the outputs and look somewhat reasonable

25.0 (100 ttbar events) has some noticeable population at high values
![all_origvssign593_ttbarwf25p0c_recojetedmreftobaseprodtofloatsassociationvector_pfcombinedcvsbjettags__reco_obj_data_](https://cloud.githubusercontent.com/assets/4676718/10063036/e27ab8fe-622e-11e5-8b01-a72bb65921d6.png)
![all_origvssign593_ttbarwf25p0c_recojetedmreftobaseprodtofloatsassociationvector_pfcombinedcvsljettags__reco_obj_data_](https://cloud.githubusercontent.com/assets/4676718/10063039/eca029a4-622e-11e5-975f-547e73c589e7.png)

and 200 events from 251721 DoubleEG: doesn't have much at high values of the discriminant
![all_origvssign593_doubleeg251721c_recojetedmreftobaseprodtofloatsassociationvector_pfcombinedcvsbjettags__reco_obj_data_](https://cloud.githubusercontent.com/assets/4676718/10063065/3874847e-622f-11e5-8b45-802b7c96beb1.png)
![all_origvssign593_doubleeg251721c_recojetedmreftobaseprodtofloatsassociationvector_pfcombinedcvsljettags__reco_obj_data_](https://cloud.githubusercontent.com/assets/4676718/10063066/3cf432c4-622f-11e5-9be1-92cc6bd7f14d.png)

mverzett commented:
@slava77, I can move the building of the vectors in the constructors in a the next PR I am preparing, but to have that pushed before tomorrow I need this one merged.

Commit:
reduce memory footprint by not storing the VPSet

Commit:
store in miniAOD candidates necesary to re-run the CTagging IVF

slava77 commented:
@monttj 
please check the changes in this PR, the AT signature is needed here
Thank you.

monttj commented:
+1
It is nice to see the first version of c-tagging!

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

gvonsem commented:
Hi @slava77 @davidlange6, could you indicate what is the time scale for this pull request to be fully merged into an integration build? I am just asking because a future PR for the DQM c-tag code we are developing will be easier to do (and to test) once the reco c-tag code of this PR is in an integration build (since the DQM code depend on the reco code).

slava77 commented:
@gvonsem 
the PR is fully signed in L2 areas.

David, if there are issues, please inform us at your earliest.

ferencek commented:
Let me also add something I asked in a private email exchange but has not been answered yet (sorry if I missed this being discussed already). In the PR a new collection of c-tag specific IVF vertices, `inclusiveCandidateSecondaryVerticesCvsL`, is introduced but not stored in the RECO/AOD event content. Was this intentional?

I'm asking because the plan was to whitelist additional tracks to have their information stored in MiniAOD. However, does this imply that `inclusiveCandidateSecondaryVerticesCvsL` will have to be remade every time MiniAOD is being produced from RECO or AOD?

slava77 commented:
On 9/25/15 11:37 AM, Dinko Ferencek wrote:
> Let me also add something I asked in a private email exchange but has
> not been answered yet (sorry if I missed this being discussed already).
> In the PR a new collection of c-tag specific IVF vertices,
> |inclusiveCandidateSecondaryVerticesCvsL|, is introduced but not stored
> in the RECO/AOD event content. Was this intentional?
> 
> I'm asking because the plan was to whitelist additional tracks to have
> their information stored in MiniAOD. However, does this imply that
> |inclusiveCandidateSecondaryVerticesCvsL| will have to be remade every
> time MiniAOD is being produced from RECO or AOD?

currently (almost?) all of the btagging is rerun when miniAOD step is
done standalone.
IMHO, it shouldn't be done by default, but it seems to be low enough
cost for miniAOD
workflow
@gpetruc can confirm


> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/11122#issuecomment-143271436>.
>

ferencek commented:
@slava77, I agree that if the b-tag discriminators produced in the standard reconstruction are used as input, then in principle there is no need to remake them. However, as long as the modules are loaded, they will be run in the unscheduled mode. There is at least one exception, though. In the re-MiniAOD of the older 74X samples, this is actually a desired behaviour.

To completely avoid such behaviour, a major restructuring of various b-tag and PAT configuration files would be needed. Note that the costliest component of the b-tagging sequence, the IVF vertex reconstruction, has been prevented from unnecessarily re-running though appropriate changes introduced in #8577.

ferencek commented on RecoBTag/Configuration/python/RecoBTag_EventContent_cff.py:
```diff
@@ -76,8 +81,10 @@
-##	'keep *_pfCombinedSecondaryVertexSoftLeptonCtagLJetTags_*_*',
-        'keep *_inclusiveCandidateSecondaryVertices_*_*'
+        'keep *_pfCombinedSecondaryVertexSoftLeptonBJetTags_*_*',
+        'keep *_inclusiveCandidateSecondaryVertices_*_*',
```
@mverzett, to keep c-tag IVF vertices, it should be sufficient to add `'keep *_inclusiveCandidateSecondaryVerticesCvsL_*_*'` after `'keep *_inclusiveCandidateSecondaryVertices_*_*'` (in 3 separate places in the cff file).

mverzett commented:
@ferencek and @slava77 I have fixed it already, it will come in the next PR that includes the whitelisting of tracks from ctagging SVs.
I will send it after dinner.

ferencek commented:
@mverzett, actually, why not just cherry-pick any additional commits and add them to this PR? I'm not sure what @slava77 prefers given that this one is fully signed already.

The point is that this next PR will be changing the event content and we need to make sure it is not too late to introduce those changes through a separate PR.

slava77 commented:
since this one is not merged yet, feel free to update.
I prefer not rebase or merges, just an incremental addition of commits.

mverzett commented:
@ferencek and @slava77 as you wish, just let me know before tomorrow afternoon.

ferencek commented:
I think we expressed our wishes :) Go ahead with cherry-picking additional commits and applying them to the branch used in this PR.

Commit:
add inclusiveCandidateSecondaryVerticesCvsL to the AOD and RECO content

Commit:
use consistent indentation (from tab to spaces)

mverzett commented:
OK, done. Everything should be OK now, but if you are not in a rush I would test it once more, to make sure I did not mess up with git cherry-picking.

cmsbuild commented:
Pull request #11122 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

@mverzett are you now done and also confident that it's all correct?

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: de5dfdbfae6c0e242cc38aa595d60aed24ddef70
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8409/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11122/8409/summary.html

mverzett commented:
@slava77 @ferencek, I think everything is in place. The only remaining thing to do is to uncomment [these two lines](https://github.com/mverzett/cmssw/blob/ctagging_76Integration/PhysicsTools/PatAlgos/python/producersLayer1/jetProducer_cfi.py#L51) once RECO RelVals with CTagging are available.

I will be offline on holidays for the next 3 weeks, but I think 99% of what was needed is done.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/11122	de5dfdb
- compared to the previously signed revision  (https://github.com/mverzett/cmssw/compare/599a3d1...de5dfdb ): inclusiveCandidateSecondaryVerticesCvsL are added to the list of stored products and their tracks to the whitelist for packed candidates to contain track info details.
- jenkins tests pass and comparisons with baseline show differences only in the miniAOD monitoring of packed candidates with stored track information (about 0.2-0.3% increase in the number of such candidates)
- additionally tested locally in CMSSW_7_6_X_2015-09-22-1500:
    - the new product inclusiveCandidateSecondaryVerticesCvsL vertex collection appears in RECO and in AOD as expected and takes up ~0.1% of AOD
    - there is no significant change in CPU
    - allocated memory is reduced by ~70 kB after the removal of the vpset copy
    - compared to the previously signed revision, there are no changes in the added ctagging variables

slava77 commented:
@monttj please check this PR
Thank you.

ferencek commented:
@mverzett, RelVals with CTagging are now available and I prepared a PR to add them to PAT tests (see #12061). Once merged, you can go ahead with the remaining updates. One thing you'll need to be careful about is the fact that CvsL IVF vertices are not stored in MiniAOD. So if someone wants to re-run c-tagging on top of MiniAOD, CvsL IVF sequence will have to be re-run. For AOD this will not be a problem. Hence, a few tweaks in jetTools.py will be needed.
